### PR TITLE
Add TypeError test in assoc_spec.rb.

### PR DESCRIPTION
**Pull request by BurdetteLamar**
Added test for an invalid name argument to ENV.assoc, which should raise TypeError.

### CODE REVIEW

File: core/env/assoc_spec.rb [hunk 1]
```diff
@@ -1,23 +1,29 @@
-require_relative '../../spec_helper'
```
Comment (eregon) [inline core/env/assoc_spec.rb]: Could you keep the `require_relative '../../spec_helper'` here?
It's convention to keep it as first line in specs.

File: core/env/env_spec_helper.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
+# Reserve variables.
```
Comment (eregon) [inline core/env/env_spec_helper.rb]: This file should be under`core/env/fixtures`, and to follow existing helper, be named `core/env/fixtures/classes.rb` and define a module ENVSpecs. No global methods should be defined in a helper file.

File: core/env/env_spec_helper.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    ENV.delete(name)
+  end
+end
+
+# Return mock object for calling #to_str.
+def mock_to_str(s)
+  mock_object = mock('name')
+  mock_object.should_receive(:to_str).and_return(s.to_s)
+  mock_object
+end
```
Comment (eregon) [inline core/env/env_spec_helper.rb]: I think this is clearer inline in each spec using it, even if it means a little bit of duplication.

File: core/env/env_spec_helper.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+require_relative '../../spec_helper'
```
Comment (eregon) [inline core/env/env_spec_helper.rb]: This should not be needed.

File: core/env/env_spec_helper.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
+# Reserve variables.
+def reserve_variables(hash)
+  hash.each_pair  do |name, value|
+    fail "Environment variable name #{name} is already in use" if ENV.include?(name)
+  end
+  @reserved_names = hash.keys
+  ENV.update(hash)
+end
```
Comment (eregon) [inline core/env/env_spec_helper.rb]: This seems over-engineered for what we need in specs.
And the spec should work, regardless of which environment variables are set.

I think a simple pattern to use is:
```ruby
before :each do
  @foo = ENV["foo"]
end

after :each do
  ENV["foo"] = @foo
end
```
This works even whether the `foo` env var is set or not.

File: core/env/assoc_spec.rb (hunk 1 above)
Comment (BurdetteLamar) [reply core/env/assoc_spec.rb]: Putting the require_relative at line 1 did not work for me.  The spec_helper launches the testing, right?  If it's cited at line 1, nothing below gets interpreted before the test runs.

Comment (eregon) [reply core/env/assoc_spec.rb]: It works for `core/array/join_spec.rb` for instance:
`$ ruby core/array/join_spec.rb`
The recommended way to run specs is to use `mspec` explicitly though:
`$ mspec core/array/join_spec.rb`

**Pull request merged**